sql: don't modify select expressions in place

RaduBerinde opened the pull request:
We currently resolve qnames in expressions, modifying them in-place. This leads
to some hacky code to correctly handle auto-transaction retries, and this scheme
won't work at all when we will split filtering expressions.

We change this by making a deep-copy of the expression before qname resolution.
If needed, in the future we could improve the visitor model to allow us to make
copies only of the nodes that change.

Changing prepared update code to look at the selectNode render target
expressions instead of relying on the in-place modification (it's actually
easier this way). @mjibson

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4418)

<!-- Reviewable:end -->

RaduBerinde commented:
An alternative to deep copying would be to change the way we use qvalues a bit: instead of replacing QualifiedName nodes in the tree, we could add a `resolved Expr`field in QualifiedName. This field would be set during qname resolution (and it wouldn't matter if it was already set before). Other than these fields, the tree wouldn't have to change for qvalue resolving. Let me know what you think of this idea and if you think it would make things cleaner in general.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAg-MsIfa18aX2oYSKL:1781144808)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I don't understand why we can ignore the expression type (i.e., ignore if it's a tuple or not). I also don't understand why we only need to look at `tableDesc.allColumnsSelector()` (`targets` when the length is taken) instead of `n.Exprs`. Could you explain that?

---

Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAgznseBA_wWjgho6WB:-1369194772)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I agree that the current setup where we mutate the expressions is problematic and error prone. On the other hand, making a full deep-copy of the expression seems excessive. I like the suggestion to make a copy of only the nodes that change. I'm a little worried that enhancing `QualifiedName` to contain a `resolved Expr` isn't a full solution as a schema change could change name resolution.

tamird commented:
Would be good to see before/after benchmark numbers.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/parser/deep_copy.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAg-phfRfdE4pRgn3vt:-KAg-phfRfdE4pRgn3vu:-426685668) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/parser/deep_copy.go#L150)):*
Can you add a test that hits this in deep_copy_test?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAhKbr_xSxAaPyIgomm:72632063)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@mjibson we are looking at all targets *except* `allColumnsSelector` - these are the  targets created from `n.Expr` in the loop above which has the exact same logic as the loop we are replacing (tuple vs dtuple vs other). Instead of going through the same loop logic again, we just look at the targets that have been created in the previous loop. I'll add a comment.

@petermattis can you expand on the problem around schema change? The `resolved Expr` inside a `QualifiedName` would be set to a `*qvalue`, and we would never reuse a previously computed value - any time we `resolveQValues` we ignore any `resolved Expr` already there. We can also have a visitor that "cleans up" (sets to nil) all `resolved` values to make sure they don't inadvertently show up.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAjUNgdlX_Nddbn4C8X:770818428)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Ok, if you're never using the `resolved Expr` between executions, that seems safe. 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAjrpChy-P1ZZDgrvQL:1529642437)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I'm not sure I understand why the prepare code needs to change?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/select_qvalue.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAkD-im_dknebBfBhdG:-KAkD-im_dknebBfBhdH:1846998525) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/select_qvalue.go#L157)):*
i'm glad we're getting rid of this code that I wasn't even convinced we needed

---

*[sql/select_qvalue.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAkCh0qmVOT8-7w8Z5C:-KAkCh0qmVOT8-7w8Z5D:-231215121) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/select_qvalue.go#L160)):*
nit: move this comment as the panic message

---

*[sql/select_qvalue.go, line 226 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAk2yfJBVkZx6y4pCkc:-KAk2yfJBVkZx6y4pCkd:1906078488) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/select_qvalue.go#L226)):*
nit: retrying transactions in general, not just "auto"

---

*[sql/update.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAk78uBbFmM3GpwoAo4:-KAk78uCQGeNJg39iyrj:-584365348) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/update.go#L96)):*
selecting all the columns here is pessimistic, right? We only need to select the cols referenced in update expressions?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAkDMnSEauFaq9UGhAA:1265179101)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
^The prepare code is relying on the input expressions being changed in-place. If you read it carefully what it does is it sets up the Select after which we can "magically" call `TypeCheck` on the same expressions..

I will give the `resolved` inside `QualifiedName` idea a try and see how it looks.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAkJznDZW3zmkutT8gH:1716374066)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I added a benchmark with a `SELECT` with a bunch of expressions (against a small table) and I don't see a regression.

I will still try the alternative approach I mentioned above to see if it makes things cleaner.

```
name                 old time/op  new time/op  delta
Select1_Cockroach-4   104µs ±10%   106µs ±34%   ~     (p=0.353 n=10+10)
Select2_Cockroach-4  1.21ms ± 8%  1.19ms ± 9%   ~       (p=0.370 n=8+9)
```

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAkXShI1ZNqOuHW7h1n:217423741)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/update.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAk78uBbFmM3GpwoAo4:-KAkY8qYyDekG6iBTIii:1818820584) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/update.go#L96)):*
I don't know this code well. At a first glance we might also need some columns to encode keys.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
By the way, I think we have a related problem with subqueries. The subquery visitor will replace subquery parts of expressions with DTuples or whatever. Upon retry of the transaction, the expression has been changed and we will not process the subqueries again. If the tables used by the subquery changed in the meantime, the results might be inconsistent.. Is this right or am I missing something?

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAkiHEoGqFVu9AL9mt0:-1142897861)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I tried the `QualifiedName.Resolved` idea and it is more hairy than I thought. There are other cases where we replace parts of expressions with `VariableExpr`s (grouping) so it doesn't fully address the problem. (by the way, these also might be problematic w.r.t retries)

Given that the benchmarks don't show any visible cost for, is this ok to merge for now?

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/deep_copy.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAg-phfRfdE4pRgn3vt:-KAl1fvC6y_3OnQuqnyk:-807610792) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/parser/deep_copy.go#L150)):*
I think we will need to implement this because a subquery might contain another subquery.
I think the current code is broken when there are subqueries and we retry anyway. I will file an issue and investigate and come up with a separate fix.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAkxokldH5V6DZ5EWeq:1768057505)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

On Wed, Feb 17, 2016 at 4:06 PM, RaduBerinde <notifications@github.com>
wrote:

> I tried the QualifiedName.Resolved idea and it is more hairy than I
> thought. There are other cases where we replace parts of expressions with
> VariableExprs (grouping) so it doesn't fully address the problem. (by the
> way, these also might be problematic w.r.t retries)
>
> Given that the benchmarks don't show any visible cost for, is this ok to
> merge for now?
> ------------------------------
>
> Review status: 0 of 4 files reviewed at latest revision, 5 unresolved
> discussions.
> ------------------------------
>
> *sql/parser/deep_copy.go, line 150 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAg-phfRfdE4pRgn3vt:-KAl1fvC6y_3OnQuqnyk:-807610792>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/parser/deep_copy.go#L150>):*
> I think we will need to implement this because a subquery might contain
> another subquery.
> I think the current code is broken when there are subqueries and we retry
> anyway. I will file an issue and investigate and come up with a separate
> fix.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAkxokldH5V6DZ5EWeq:1768057505>*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4418#issuecomment-185403184>
> .
>

tamird commented:
LGTM

---

Reviewed 1 of 4 files at r4, 2 of 3 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/deep_copy.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAg-phfRfdE4pRgn3vt:-KAl3LHHgTinUl8spM-d:-901912962) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/parser/deep_copy.go#L150)):*
I'm not sure what you mean - I was only suggesting you add a test that documents that `DeepCopy()` on a `Subquery` is a no-op.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAl3do68r59Czr4zv3r:676834233)*
<!-- Sent from Reviewable.io -->

mjibson commented:
LGTM

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-:-KAl6PTNQfYQbL6EsXFQ:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/parser/deep_copy.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAg-phfRfdE4pRgn3vt:-KAl8eUCIbe1DUJBSi5Q:2037905616) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6efa1c2a5ae059613cdbe4a121c579b78d1451/sql/parser/deep_copy.go#L150)):*
I know, but I realized that we will actually need to implement this to make a full copy of `Subquery` as well, along with more fixes related to retrying selects with subqueries.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/parser/deep_copy.go, line 206 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAlFE6N_0PCaw0BQrtq:-KAlFE6N_0PCaw0BQrtr:-442538161) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d38f683dede2850f59e88f3b508b5acc1f3f72d/sql/parser/deep_copy.go#L206)):*
I should have caught this earlier, but @dt just brought this up to me. This won't correctly perform a deep copy on the `DDecimal`. The problem is that the underlying `big.Int` contains a `big.nat`, which is a slice of `Word`s. If we want to perform this, we need to make sure that the `big.Int` is properly deep copied itself.

I believe that would look something like:
```
func (t *DDecimal) DeepCopy() Expr {
    dd := &DDecimal{}
    dd.UnscaledBig().SetBytes(t.UnscaledBig().Bytes())
    dd.SetScale(t.Scale())
    return dd
}
```

That said, I believe all of our builtin functions and eval ops treat the decimal args as immutable, so this probably isnt critical.

---

*[sql/parser/deep_copy.go, line 224 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAlGKC6tEH26G79OZ7n:-KAlGKC7_Fxgeh9WYPxa:1689476734) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d38f683dede2850f59e88f3b508b5acc1f3f72d/sql/parser/deep_copy.go#L224)):*
Since I'm here, I would expect this to need to allocate a new `DTuple` and recursively call `DeepCopy` on each element.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/parser/deep_copy.go, line 206 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418#-KAlFE6N_0PCaw0BQrtq:-KAlKZ0PDmBfdrauLNdU:760013714) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d38f683dede2850f59e88f3b508b5acc1f3f72d/sql/parser/deep_copy.go#L206)):*
Thanks, yeah, it's actually not a complete deep copy - the leaves don't necessarily get copied, and it's ok because we never modify them in place. We definitely need a more consistent model though. I am working on improving this.. I am considering changing the deep copy stuff with a "node copy" routine and changing the visitor model to always create new nodes whenever something is changed (that way all nodes become effectively immutable).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4418)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: select benchmark

Adding a benchmark for a SELECT statement with non-trivial target and filter
expressions.

RaduBerinde committed:
sql: don't modify select expressions in place

We currently modify expressions in-place when we resolve qnames. This leads to
some hacks to correctly handle auto-transaction retries, and it won't work at
all when we will split filtering expressions.

We change this by making a deep-copy of the expression before qname resolution.
If needed, in the future we could improve the visitor model to allow us to make
copies only of the nodes that change.

Changing prepared update code to look at the selectNode render target
expressions instead of relying on the in-place modification (it's actually
easier this way).
